annotated.go

```diff
@@ -637,62 +1278,110 @@ func (ann *annotated) Build() (interface{}, error) {
 		return nil, fmt.Errorf("invalid annotation function %T: %w", ann.Target, err)
 	}
 
-	resultTypes, remapResults, err := ann.results()
-	if err != nil {
-		return nil, err
+	// app
+	ann.prebuild()
+
+	var err error
+	var lcHookAnns []*lifecycleHookAnnotation
+	for _, annotation := range ann.Annotations {
+		switch annotation := annotation.(type) {
+		case *lifecycleHookAnnotation:
+			lcHookAnns = append(lcHookAnns, annotation)
+		default:
+			if ann.Target, err = annotation.build(ann); err != nil {
+				return nil, err
+			}
+		}
 	}
-	paramTypes, remapParams, hookParams := ann.parameters(resultTypes...)
 
-	hookFns := make([]reflect.Value, len(ann.Hooks))
-	for i, builder := range ann.Hooks {
-		if hookFn := builder.Build(resultTypes...); !hookFn.IsZero() {
-			hookFns[i] = hookFn
+	ann.postbuild()
+
+	for _, la := range lcHookAnns {
+		if ann.Target, err = la.build(ann); err != nil {
+			return nil, err
```

Comment:
So I checked the implementation of `*lifecycleHookAnnotation.build()`, and apparently it's impossible to return non-nil error. This is because I do the dependency check for the hook function by invoking a function in a private scope when the annotated function is called. So if there's a problem with dependencies of the hook function, it fails while the app/module is running provides, instead of failing while building the annotated function. There's no other failure case because all other sanity checks on hook functions when `Annotation.apply()` is called.

Since I do not want the signature of build() method of Annotation interface to change because other types of annotation may return error when building, should I just have an unhandled error here?

Also a similar case of uncovered code is the call to `scope.Provide` in the hookInstaller. I know this call to Provide cannot fail because I have the full control over what's passed to that Provide, and I have codes to make sure we don't provide an invalid constructor to that scope. (by removing context.Context and deduplicating types)